File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-2.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @param {number} n
     * @return {ListNode}
     */
    removeNthFromEnd(head, n) {
        //approac 1 
        // if(!head) return head;
        // let len=0,curr=head;
        // while(curr){
        //     len++;
        //     curr=curr.next;
        // }
        // let index=len-n;
        // if(index==0) return head.next;
        // curr=head;
        // for(let i=1;i<index;i++){
        //     curr=curr.next;
        // }
        // curr.next=curr.next.next || null;
        // return head;

        //solution 2
        if(!head) return head;
        let slow=head,fast=head;
        while(n>0){
            fast=fast.next;
            n--;
        }
        if(!fast) return head.next;
        while(fast && fast.next){
            slow=slow.next;
            fast=fast.next;
        }
        slow.next=slow.next.next;
        return head;
    }
}

File: Data Structures & Algorithms/add-two-numbers/submission-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if(l1==null) return l2;
        if(l2==null) return l1;

        ListNode curr1=l1, curr2=l2;
        ListNode ansList = new ListNode(-1), temp=ansList;
        int carry=0;
        while(curr1!=null || curr2!=null || carry>0){
            int sum = carry;
            if(curr1!=null){
                sum+=curr1.val;
                curr1=curr1.next;
            }
            if(curr2!=null){
                sum+=curr2.val;
                curr2=curr2.next;
            }
            temp.next=new ListNode(sum%10);
            temp = temp.next;
            carry = sum/10;
        }
        return ansList.next;
    }
}
